protocol/test-sequencing.md

```diff
@@ -0,0 +1,216 @@
+# Summary
+
+Interop testing requires completion of a long
+[checklist](https://github.com/ethereum-optimism/optimism/issues/13855) of particularly constructed blocks and reorgs.
+To implement these tests, we need new tooling that can include transactions and perform reorgs in exact ways.
+
+While implementing the test tooling, we have an opportunity to prepare separation of sequencing from the op-node,
+to reduce op-node complexity and encapsulate sequencing concerns better.
+
+# Problem Statement + Context
+
+Sequencing in the op-stack can be bucketed into different types:
+- Legacy sequencing (tx-pool, op-node with op-geth, recommit, priority fee)
+- Flashblocks (rollup-boost proxy between op-node and op-geth, mux to builder node, 250ms incremental block building)
+- Testing (action-tests, next-up interop work)
+
+The testing type of work is in need of change, to provide the level of block-building control
+that [action-tests](https://github.com/ethereum-optimism/optimism/tree/develop/op-e2e/actions)
+have to fully end-to-end tests like those backed by
+[Kurtosis](https://github.com/ethereum-optimism/optimism/tree/develop/kurtosis-devnet).
+
+Instead of building a limited testing tool, we can consider building a full sequencer service:
+- Op-node complexity can be reduced.
+- Separate sequencing from verifying, to not unnecessarily disrupt sequencing work.
+  This is a known issue seen by chains who have heavy / spiking verification workloads.
+- Block-building in op-geth is not a good starting point.
+    - On every tx, geth [hashes the full block-header again, unnecessarily](https://github.com/ethereum-optimism/op-geth/pull/462).
+    - Block-building does not prepare a cache for block-processing.
+    - Block recommit code is complicated, wrangled into the geth block-building `miner` package.
+      Regardless of the event-control we add to op-node, this legacy geth code makes testing impractical.
+
+# Proposed Solution
+
+This is an idea to separate the sequencing from op-node into its own service with test-RPC.
+We could call this service `op-sequencer`, and maybe use a different name while
+it's only servicing testing and is still more experimental.
+
+For testing and experiments, the service should build blocks manually, with full control over reorgs/inclusion/ordering.
+
+For regular operation, and rbuilder-backed flashblocks,
+we can make the service fall back to only run the block-proposing part of sequencing: see "engine mode" section.
+
+## Increments
+
+### Testing scope
+
+Test-functionality: Build the block without the engine API.
+
+The test-functionality is inspired a lot by the action-tests:
+having exact control over block-construction is very useful for testing.
+And by putting it in a service, we can use it more places than action-tests, like Kurtosis / devnets.
```

Comment:
I understand there are other use-cases that justify the service approach, but for test-only cases, couldn't we also get away with just a library providing the block building capability?
I might be missing something

Reply:
A library can produce the block, but then the block still needs to be inserted into the canonical chain, including the act of reorging.